File: python/14.longest-common-prefix.py
#
# @lc app=leetcode id=14 lang=python3
#
# [14] Longest Common Prefix
#

# @lc code=start
from collections import deque
from typing import List


class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        prefix = ""
        minlen = min([len(string) for string in strs])
        for i in range(minlen):
            strsubs = [st[i] for st in strs]
            if len(set(strsubs)) == 1:
                prefix += strs[0][i]
            else:
                break
        return prefix

            
        
        
# @lc code=end

File: rust/14.longest-common-prefix.rs
/*
 * @lc app=leetcode id=14 lang=rust
 *
 * [14] Longest Common Prefix
 */

// @lc code=start
impl Solution {
    pub fn longest_common_prefix(strs: Vec<String>) -> String {
        let min_length = strs.iter().map(|s| s.len()).min().unwrap_or(0);
        let mut prefix = String::new();
        for i in 0..min_length {
            let c = strs.get(0).map(|x| x.chars().nth(i)).flatten();
            match c {
                Some(c) => match strs.iter().all(|s| s.chars().nth(i) == Some(c)) {
                    true => prefix.push(c),
                    false => break,
                },
                None => break,
            }
        }
        prefix
    }
}
// @lc code=end
